12-offer-encoding.md

```diff
@@ -0,0 +1,1051 @@
+# BOLT #12: Flexible Protocol for Lightning Payments
+
+# Table of Contents
+
+  * [Limitations of BOLT 11](#limitations-of-bolt-11)
+  * [Payment Flow Scenarios](#payment-flow-scenarios)
+  * [Encoding](#encoding)
+  * [TLV Fields](#tlv-fields)
+  * [Invoices](#invoices)
+  * [Offers](#offers)
+  * [Invoice Requests](#invoice-requests)
+  * [Invoice Errors](#invoice-errors)
+
+# Limitations of BOLT 11
+
+The BOLT 11 invoice format has proven popular, but has several
+limitations:
+
+1. The entangling of bech32 encoding makes it awkward to send
+   in other forms (i.e. inside the lightning network itself).
+2. The signature applying to the entire invoice makes it impossible
+   to prove an invoice without revealing its entirety.
+3. Fields cannot generally be extracted for external use: the `h`
+   field was a boutique extraction of the `d` field, only.
+4. The lack of 'it's OK to be odd' rule makes backwards compatibility
+   harder.
+5. The 'human-readable' idea of separating amounts proved fraught:
+   `p` was often mishandled, and amounts in pico-bitcoin are harder
+   than the modern satoshi-based counting.
+6. The bech32 encoding was found to have an issue with extensions,
+   which means we want to replace or discard it anyway.
+7. The `payment_secret` designed to prevent probing by other nodes in
+   the path was only useful if the invoice remained private between the
+   payer and payee.
+8. Invoices must be given per-user, and are actively dangerous if two
+   payment attempts are made for the same user.
+
+
+# Payment Flow Scenarios
+
+Here we use "user" as shorthand for the individual user's lightning
+node, and "merchant" as the shorthand for the node of someone who is
+selling or has sold something.
+
+There are two basic payment flows supported by BOLT 12:
+
+The general user-pays-merchant flow is:
```

Comment:
I was playing around with offers and I have the feeling this flow should be automated (similar to the `commitment_signed` and `revoke_and_ack` handshakes that we do to move forward the state and hide from the user). I have to admit that at first I was a little bit hesitant to post that here as this could ultimately be an implementation level choice but I decided to leave the comment here since I think that from a user experience perspective (in particular with recurring offers) the flow should be as simple as: 

1. Merchant Makes an offer
2. User accept the offer (or subscribes to it or what ever you call it)
3. (optionally): User unsubscribe from it / merchant withdraws the offer. 
 
My rational: If you were for example to make an offer that I would pay you 100 sats every hour I would not want to fetch and pay an invoice every hour. The low level data flow of fetching and paying invoices should be going on as soon as a user accepts to the offer. As subscription models live from being automated and without bothering the user. 

Of course this automated idea triggers some consequences. E.g. what happens if the user who subscribed to an recurring offer lacks temporarily the liquidity to pay the invoice. Should the Node pay the invoice as soon as liquidity becomes available or shout the user be presented the invoice and told to now pay it manually? 

Of course some users wish to have control. So Maybe there could be an automated flag either in the make offer API or in the accept offer (seems most natural that a merchant suggests automated and the user would then decide and potentially accept the offer without automation)

Reply:
There's an easy extension to allow invoices to be generated without re-requesting offers.  It's perfectly backwards compatible so I didn't add it to bolt 12 yet:

1. A new (odd) tlv entry which says how many times the invoice can be reused.
2. The server performs SHA256() on the payment_preimage this many times before presenting it.
3. Now when you get one preimage, that's the payment_hash for the next.
4. If client does not understand this, no problems: they just request the new invoice and it happens to have the same payment_hash as the last payment_preimage.

This works with payment_points too, ofc.

Reply:
"without re-requesting offers" - I assume you meant "without re-requesting _invoices_"?